src/build_system.ml

```diff
@@ -617,18 +645,18 @@ let () =
     pending_targets := Path.Set.empty;
     Path.Set.iter fns ~f:Path.unlink_no_err)
 
-let clear_targets_digests_after_rule_execution targets =
-  let missing =
-    List.fold_left targets ~init:Path.Set.empty ~f:(fun acc fn ->
-      match Unix.lstat (Path.to_string fn) with
-      | exception _ -> Path.Set.add acc fn
-      | (_ : Unix.stats) ->
-        Utils.Cached_digest.remove fn;
-        acc)
+let compute_targets_digest_after_rule_execution targets =
+  let good, bad =
+    List.partition_map targets ~f:(fun fn ->
+      match Utils.Cached_digest.refresh fn with
+      | digest -> Left digest
+      | exception (Unix.Unix_error _ | Sys_error _) -> Right fn)
```

Comment:
It might be nice to make `refresh` and `file` to return a `Result` type. I see that we're handling this exception in two cases.

Reply:
I'm not sure it improves things much given that in these cases as we just ignore the exception

Reply:
I think what it improves is not having to remember to catch 2 exceptions. But I won't press this.